### PR TITLE
cmake: Add support for add_custom_target() and add_dependencies()

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

File: mesonbuild/cmake/traceparser.py [hunk 1]
```diff
@@ -479,7 +497,10 @@ def _lex_trace(self, trace):
             func = mo_file_line.group(4)
             args = mo_file_line.group(5)
             args = parse_generator_expressions(args)
-            args = args.split(' ')
+            try:
+                args = shlex.split(args)
+            except:
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [B901] blind except: statement ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637200))

<sub>posted by Sider</sub>

Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [B001] Do not use bare `except:`, it also catches unexpected events like memory errors, interrupts, system exit, and so on.  Prefer `except Exception:`.  If you're sure what you're doing, be explicit and write `except BaseException:`. ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637199))

<sub>posted by Sider</sub>

**Comment by mensinda** [comment]
When translating the dependency chain to meson, you should make sure that the `add_custom_target` is run before *any* code of the dependent target is compiled. As far as I know, meson will only do this if in the output of a meson `custom_target` is at least one header file. So you may have to generate a dummy header file that is then "consumed" in the build target.

**Comment by xclaesse** [comment]
Yes, that's what it does. When the custom target has no output, it create a dummy one, that's why I changed run_ctgt.py.

The generated code look like this:
```meson
custom_tgt_generate_exports = custom_target(
  'custom_tgt_generate_exports',
  input : [],
  output : ['custom_tgt_generate_exports'],
  ...
```

An empty file `custom_tgt_generate_exports` will be created and used as dependency. It is passed as source file to build targets, and meson ignore it because it doesn't have a known compiler extention.

**Comment by mensinda** [comment]
My point was that it should *always* be a `*.h` file to create a build dependency, not a link dependency. If it isn't a `*.h` file, it might happen that the custom target is executed *while* other parts of the project are built. This can lead to missing headers at build time.

At least that was the case when I made my original PR, but I doubt that this has changed.

File: mesonbuild/cmake/traceparser.py [hunk 2]
```diff
@@ -396,6 +404,55 @@ def _cmake_set_target_properties(self, tline: CMakeTraceLine) -> None:
...
+            for i in arg.split(';'):
+                target = self.targets.get(i)
+                if not target:
+                    return self._gen_exception('install', 'target not found', tline)
+                target.install = True
+                install_targets.append(target)
+
+        def handle_destination(arg: str):
+            install_dir = arg
+            fn = None
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [F841] local variable 'fn' is assigned to but never used ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637210))

<sub>posted by Sider</sub>

File: mesonbuild/cmake/traceparser.py [hunk 3]
```diff
@@ -396,6 +404,55 @@ def _cmake_set_target_properties(self, tline: CMakeTraceLine) -> None:
...
+        # DOC: https://cmake.org/cmake/help/latest/command/install.html
+        args = list(tline.args)
+
+        if len(args) < 2:
+            return self._gen_exception('install', 'takes at least 2 arguments', tline)
+
+        fn = None
+        install_targets = []
+        install_dir = None
+        def handle_target(arg: str):
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [E306] expected 1 blank line before a nested definition, found 0 ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637208))

<sub>posted by Sider</sub>

File: mesonbuild/cmake/traceparser.py [hunk 4]
```diff
@@ -396,6 +404,55 @@ def _cmake_set_target_properties(self, tline: CMakeTraceLine) -> None:
...
+            for i in arg.split(';'):
+                target = self.targets.get(i)
+                if not target:
+                    return self._gen_exception('install', 'target not found', tline)
+                target.install = True
+                install_targets.append(target)
+
+        def handle_destination(arg: str):
+            install_dir = arg
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [F841] local variable 'install_dir' is assigned to but never used ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637209))

<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
Hmm, if the command writes a header file, it should be listed in artifacts, and that will be picked already with `isGenerated` flag in json. In this case the custom target has no output, so I don't think it can be considered a compile time dependency. In my case it generates version script passed to linker.

**Comment by mensinda** [comment]
In an ideal world, yes it would. However, we are dealing with CMake here :)

It is possible to never specify the `BYPRODUCTS` if you are manually using `add_dependency`. Thus it is valid CMake to have a custom target with no *declared* byproducts, that generates header files, used in other build targets.

In this case, we have to assume, for every `add_custom_target`, that there might be a header file in the output that the user hasn't bothered to specify. This isn't the case for `add_custom_command` because there is no target name and all the outputs **have to** be specified by the user to create the dependency chain.

With your PR, it should be enough to just append a `.h` to the output and everything (should) work.

File: mesonbuild/cmake/traceparser.py [hunk 5]
```diff
@@ -479,7 +538,10 @@ def _lex_trace(self, trace):
             func = mo_file_line.group(4)
             args = mo_file_line.group(5)
             args = parse_generator_expressions(args)
-            args = args.split(' ')
+            try:
+                args = shlex.split(args)
+            except:
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [B901] blind except: statement ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637212))

<sub>posted by Sider</sub>

Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [B001] Do not use bare `except:`, it also catches unexpected events like memory errors, interrupts, system exit, and so on.  Prefer `except Exception:`.  If you're sure what you're doing, be explicit and write `except BaseException:`. ([link](https://sider.review/gh/repos/19784232/pulls/5989#issue-4637211))

<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
> With your PR, it should be enough to just append a `.h` to the output and everything (should) work.

Ok, will do that then. Hackish, but as you said, we are dealing with cmake after all...

**Comment by xclaesse** [comment]
> > With your PR, it should be enough to just append a `.h` to the output and everything (should) work.
> 
> Ok, will do that then. Hackish, but as you said, we are dealing with cmake after all...

done :-)

**Comment by xclaesse** [comment]
@mensinda ah, CI finally pass here. Can you review please?

**Comment by mensinda** [comment]
Ping

**Comment by mensinda** [comment]
@xclaesse gave me permission on IRC to fix the conflicts and do some minor fixes.

So, once the CI is green, this PR is ready to merge.

**Comment by mensinda** [comment]
The OSX error is unrelated.

**Comment by mensinda** [comment]
The CI is finally green. This is ready to merge from my part.

**Pull request merged**